### PR TITLE
Rate-limit new prebuilds to 50 / user / minute to prevent incidents

### PR DESCRIPTION
**Pull request by jankeromnes**
## Description

Rate-limit new prebuilds to 50 / user / minute to prevent incidents

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/5176

## How to test

1. Fork https://gitlab.com/jankeromnes/test-gitpod-prebuilds-rate-limit
2. Add your fork as a project here: https://jx-rt-lmt-prblds.staging.gitpod-dev.com/new?team=user
3. Open your fork in Gitpod (any Gitpod) and run `./push-100.sh`
4. Once all 100 commits have been pushed (20 commits to 5 branches), verify that:
    - most prebuilds actually started (but not all 100)
    - some prebuilds never started but got cancelled instead (between 0 and 50, depending on when the burst happened in the 1min rate-limit window)
    - `kubectl logs server-[TAB] server | grep -i "rate limit"` contains some rate limit warnings (as many as the cancelled prebuilds)

## Release Notes

```release-note
Rate-limit new prebuilds to 50 / user / minute to prevent incidents
```

### CODE REVIEW

**Comment by jankeromnes** [comment]
Re-roll 🎲

/werft run

**Comment by jankeromnes** [comment]
**Expected**
- 100 commits pushed (20 to 5 branches)
- 50 prebuilds started
- 50 rate limit warnings

**Current state**
- 100 commits pushed (20 to 5 branches)
- 100 prebuilds started
- 13 rate limit warnings

Something seems off here 🤔 (do GitLab webhooks auto-retry on rate limit? 😳)

**Comment by gtsiolis** [comment]
See also [relevant discussion](https://gitpod.slack.com/archives/C02EN94AEPL/p1646081001709509) (internal) in case this is related. Cc @jankeromnes @JanKoehnlein @jldec

**Comment by jankeromnes** [comment]
Ah, cool, I didn't know we could see webhook errors on GitLab. Thanks @gtsiolis!

Would be curious to see if we can get the actual error messages when a webhook call fails.

**Comment by gtsiolis** [comment]
> Would be curious to see if we can get the actual error messages when a webhook call fails.

@jankeromnes Yes, we can. Here're the two most recent errors seen in our fork.
1. https://gitlab.com/gitpod-io/gitlab/-/hooks/10823912/hook_logs/2455346235 (internal)
2. https://gitlab.com/gitpod-io/gitlab/-/hooks/10823912/hook_logs/2455045008 (internal)

**Comment by jankeromnes** [comment]
Hey, thanks a lot for surfacing these. 💯🙏

Both seem to have failed with:

```
Internal error occurred while delivering this webhook.
Error: Net::ReadTimeout
```

I wonder if the timeout means Gitpod didn't respond in time, or if it's something inside GitLab that timed out. 🤔

In any case, this feature looks super useful for debugging webhook problems (especially now that we're inventing a new webhook problem: the rate limit 😅)

**Comment by jankeromnes** [comment]
Two new errors on the current deployment, and rate-limiting doesn't work 🙄

```
TypeError: Cannot read properties of undefined (reading 'statusCode')
    at /app/node_modules/@gitbeaker/node/dist/index.js:168:43
    at step (/app/node_modules/@gitbeaker/node/dist/index.js:82:23)
    at Object.next (/app/node_modules/@gitbeaker/node/dist/index.js:63:53)
    at fulfilled (/app/node_modules/@gitbeaker/node/dist/index.js:53:58)
```

```
(node:29) UnhandledPromiseRejectionWarning: Error: NotFoundError
    at Object.<anonymous> (/app/node_modules/@gitpod/server/dist/src/errors/index.js:26:41)
    at Generator.next (<anonymous>)
    at /app/node_modules/@gitpod/server/dist/src/errors/index.js:13:71
    at new Promise (<anonymous>)
    at __awaiter (/app/node_modules/@gitpod/server/dist/src/errors/index.js:9:12)
    at Object.create (/app/node_modules/@gitpod/server/dist/src/errors/index.js:21:16)
    at GitlabContextParser.<anonymous> (/app/node_modules/@gitpod/server/dist/src/gitlab/gitlab-context-parser.js:187:52)
    at Generator.next (<anonymous>)
    at fulfilled (/app/node_modules/@gitpod/server/dist/src/gitlab/gitlab-context-parser.js:19:58)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

File: components/server/src/websocket/websocket-connection-manager.ts [hunk 1]
```diff
@@ -358,7 +358,7 @@ class GitpodJsonRpcProxyFactory<T extends object> extends JsonRpcProxyFactory<T>
                     throw rlRejected;
                 }
                 log.warn({ userId }, "Rate limiter prevents accessing method due to too many requests.", rlRejected, { method });
-                throw new ResponseError<RateLimiterError>(ErrorCodes.TOO_MANY_REQUESTS, "too many requests", { method, retryAfter: Math.round(rlRejected.msBeforeNext / 1000) || 1 });
+                throw new ResponseError<RateLimiterError>(ErrorCodes.TOO_MANY_REQUESTS, "too many requests", { method, retryAfter: Math.ceil(rlRejected.msBeforeNext / 1000) || 1 });
```
Comment (geropl) [inline components/server/src/websocket/websocket-connection-manager.ts:361]: :+1:

**Comment by geropl** [comment]
@jankeromnes Thx for tackling this long-standing yet very annoying issue. Two thoughts:
 - the current (draft) version throttles per user. With an additional method [here](https://github.com/gitpod-io/gitpod/blob/35e97c10643734396c4368240d453ad734c82bdf/components/server/src/auth/rate-limiter.ts#L253-L255) we could turn it into per-repo
 - have you thought about using the DB as ground-truth about how many workspace are already running per repo? :thinking: This would:
   - effectively share rate-limiting state across `server` instances: the current approach allows for 6 x 50 = 300 prebuilds / 1 minutes, potentially
   - this could pave the way to support user-friendlier modes of handling over-traffic compared to "accept first, drop all others"

**Comment by jankeromnes** [comment]
> > effectively share rate-limiting state across `server` instances: the current approach allows for 6 x 50 = 300 prebuilds / 1 minutes, potentially
> 
> Aha, right, I'm currently testing with a single `server` instance, which is why I didn't observe this.

I've tested this PR with 6 servers, using `kubectl scale deployments/server --replicas=6`.

```
mysql> select state, count(id) prebuilds from d_b_prebuilt_workspace where creationTime > '2022-03-02 14:42' group by state order by
prebuilds desc;
+-----------+-----------+
| state     | prebuilds |
+-----------+-----------+
| available |       100 |
+-----------+-----------+
1 row in set (0.01 sec)
```

```
mysql> select state, count(id) prebuilds from d_b_prebuilt_workspace where creationTime > '2022-03-02 14:46' group by state order by
prebuilds desc;
+-----------+-----------+
| state     | prebuilds |
+-----------+-----------+
| available |       100 |
+-----------+-----------+
1 row in set (0.01 sec)
```

Looks like this defeats the rate-limiter. 😞

**Comment by jankeromnes** [comment]
While this approach (i.e. using the server rate-limiter inside the PrebuildManager) basically works, it has the problem of multiplying the allowed rate limit by the number of server instances in the cluster.

I still believe that we could fix the rate-limiter itself to use the DB in order to share a single state between all server instances of a cluster (thus fixing this approach, and also fixing every other rate-limited server method).

However, for the specific issue of prebuild spam, we've [discussed](https://github.com/gitpod-io/gitpod/issues/5176#issuecomment-1057906713) another approach (i.e. not using the rate-limiter, but making the PrebuildManager check the DB for how many prebuilds were previously started on the same cloneURL), which seems even simpler and could immediately prevent this type of outage.

Thus, closing this Pull Request for now. @andrew-farries and @easyCZ, if you do find anything useful in this PR, please feel free to recycle it as you see fit. 🙂 Many thanks for taking this over! I'm also happy to help with any questions here.

**Pull request closed without merging**